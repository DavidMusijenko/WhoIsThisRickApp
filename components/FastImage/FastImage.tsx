import { StyleSheet, FlatList } from 'react-native';
import { View, Text } from '../Themed';
import { useNavigation } from '@react-navigation/native';
import { useInfiniteQuery } from '@tanstack/react-query';
import { fetchAllCharacters } from '../../api/RickApi';
import { Spinner } from 'native-base';
import { ImageEl } from './ImageEl/ImageEl';
import { Dimensions } from 'react-native';

const FastImage = () => {
  const navigation = useNavigation();
  const windowWidth = Dimensions.get('window').width;
  const {
    data,
    isLoading,
    hasNextPage,
    fetchNextPage,
    isFetchingNextPage,
  }: any = useInfiniteQuery(
    ['characters'],
    async ({ pageParam }) => await fetchAllCharacters({ pageParam }),

    {
      getNextPageParam: (lastPage) => {
        if (lastPage.info?.next !== null) {
          return lastPage.info?.next;
        } else {
          return null;
        }
      },
    }
  );

  console.log(data);

  const renderItem = ({ item }: any) => (
    <ImageEl
      navigation={navigation}
      image={item.image}
      id={item.id}
      name={item.name}
      windowWidth={windowWidth}
    />
  );

  const itemExtractorKey = (item: any, index: number) => {
    return index.toString();
  };

  const loadMore = () => {
    if (hasNextPage) {
      fetchNextPage();
    }
  };

  const renderSpinner = () => {
    return <Spinner size='lg' />;
  };

  return isLoading ? (
    <View style={styles.container}>
      <Spinner size='lg' />
    </View>
  ) : (
    <View>
      <FlatList
        data={data?.pages?.map((page: any) => page.results).flat()}
        renderItem={renderItem}
        keyExtractor={itemExtractorKey}
        onEndReached={loadMore}
        onEndReachedThreshold={0.3}
        ListFooterComponent={isFetchingNextPage ? renderSpinner : null}
        numColumns={3}
      />
    </View>
  );
};

export default FastImage;

const styles = StyleSheet.create({
  container: {
    marginTop: 50,
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
