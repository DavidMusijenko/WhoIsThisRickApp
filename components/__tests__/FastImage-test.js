import FastImage from 'react-native-fast-image';
import renderer from 'react-test-renderer';

it(`renders correctly`, () => {
  const tree = renderer.create(<FastImage />).toJSON();

  expect(tree).toMatchSnapshot();
});
