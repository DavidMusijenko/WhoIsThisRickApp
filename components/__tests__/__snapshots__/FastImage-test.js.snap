// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <FastImageView
    defaultSource={null}
    resizeMode="cover"
    source={null}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;
